### PR TITLE
Distributed Evaluation

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Allow for sped up evaluation via distributed evaluation.

Involves a refactor of distributed.

**Testing steps**
Manual testing. New CI.

### CODE REVIEW

**Comment by emilydinan** [comment]
CC @wyshi

**Comment by stephenroller** [comment]
Ready for review. It looks like a bigger diff than it is, mostly because of a refactor.

File: parlai/scripts/multiprocessing_eval.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+
+"""
+Main launch script for single-host, multi-GPU training.
+
+This is a drop-in replacement for train_model.py.  This script will launch N
+subprocess, each which runs the full training loop independently.
+
+Uses torch.nn.parallel.DistributedDataParallel for its main uses.  Agents must
+specifically implement the wrapper of DistributedDatParallel, but all
```
Comment (EricMichaelSmith) [inline parlai/scripts/multiprocessing_eval.py:15]: nano-nit: typo in Data

File: parlai/scripts/multiprocessing_eval.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import os
+import signal
+import parlai.utils.distributed as distributed_utils
+import parlai.scripts.eval_model as eval_model
+
+
+def multiprocess_eval(
+    rank, opt, port=61337, rank_offset=0, gpu=None, hostname='localhost'
+):
+    with distributed_utils.distributed_context(
```
Comment (EricMichaelSmith) [inline parlai/scripts/multiprocessing_eval.py:30]: Maybe a docstring here? Or explain multiprocess_eval in the main module docstring above, which only focuses on discussing training and not eval?

File: parlai/utils/distributed.py [hunk 3]
```diff
@@ -285,3 +289,108 @@ def sync_parameters(model: torch.nn.Module) -> bool:
...
+def distributed_context(
+    rank, opt, port=61337, rank_offset=0, gpu=None, hostname='localhost'
+):
+    """
+    Subprocess which initializes distributed training, and begins training.
+
+    This should be launched n times for n GPUs; this is handled either in main
+    or via srun.
+
+    :param int rank: This process's rank - 1. (Starts at -1 ... n - 2). See comments.
```
Comment (EricMichaelSmith) [inline parlai/utils/distributed.py:304]: Which comment is the -1 discussed in?

Comment (stephenroller) [reply parlai/utils/distributed.py:304]: Good catch.

**Pull request merged**